### PR TITLE
LTR refactoring with spacy upgraded and index added to pre-built

### PR DESCRIPTION
**Pull request by stephaniewhoo**
+ space upgrade to 3.0.6 for preprocessing
+ index added to pre-built

### CODE REVIEW

**Comment by stephaniewhoo** [comment]
@yuki617 I have done walking through steps once, can you walk through it when you have time to see if we can get same number? 
Thanks

File: docs/experiments-ltr-msmarco-passage.md [hunk 1]
```diff
@@ -28,12 +28,12 @@ The above scripts convert queries to json objects with `text`, `text_unlemm`, `r
 The first two scripts take ~1 min and the third one is a bit longer (~1.5h).
 
 ```bash
-wget https://www.dropbox.com/s/se5kokw1qqu8yxs/lucene-index-msmarco-passage-ltr.tar.gz -P indexes/ 
-tar -xzvf indexes/lucene-index-msmarco-passage-ltr.tar.gz 
+wget https://www.dropbox.com/s/2h22kt43fcnngxg/index-msmarco-passage-ltr-20210519-e25e33f.tar.gz -P indexes/
```
Comment (lintool) [inline docs/experiments-ltr-msmarco-passage.md:31]: with pre-built index, you don't need explicit downloading...

You can do something like:
```
searcher = SimpleSearcher.from_prebuilt_index('msmarco-passage')
```

Directly from your script?

File: docs/experiments-ltr-msmarco-passage.md [hunk 2]
```diff
@@ -72,7 +73,7 @@ After the run finishes, we can evaluate the results using the official MS MARCO
 $ python tools/scripts/msmarco/msmarco_passage_eval.py \
    tools/topics-and-qrels/qrels.msmarco-passage.dev-subset.txt runs/run.ltr.msmarco-passage.tsv
 #####################
-MRR @10: 0.24819336426069846
+MRR @10: 0.24709612498294367
```
Comment (lintool) [inline docs/experiments-ltr-msmarco-passage.md:76]: Any idea why the effectiveness went down? Might we need to train the model with the new spacy tokenizations, etc.?

File: docs/experiments-ltr-msmarco-passage.md (hunk 1 above)
Comment (stephaniewhoo) [reply docs/experiments-ltr-msmarco-passage.md:31]: hrm, actually we are using index sth like `fe = FeatureExtractor('index-msmarco-passage-ltr-20210519-e25e33f.tar.gz', max(multiprocessing.cpu_count()//2, 1))`
and FeatureExtractor is a class we defined in Anserini... 
we can use pre-built maybe, might need to reorganize java side in a large scale

File: docs/experiments-ltr-msmarco-passage.md (hunk 2 above)
Comment (stephaniewhoo) [reply docs/experiments-ltr-msmarco-passage.md:76]: I also feel weird. But this is the result I get by redo all the steps with space 3.0.6 (preprocess, training and reranking)

File: docs/experiments-ltr-msmarco-passage.md (hunk 1 above)
Comment (lintool) [reply docs/experiments-ltr-msmarco-passage.md:31]: Oh, I see.

How about doing this as a one-liner:

```
python -c "from pyserini.search import SimpleSearcher; SimpleSearcher.from_prebuilt_index('xxxxx')"
```

And then pass the `~/.cache/....` path into Pyserini.

That way, you don't need to worry about the Dropbox path changing again in the future...

Comment (stephaniewhoo) [reply docs/experiments-ltr-msmarco-passage.md:31]: ah, I see your point. 
yea, that sg.
 next time we change the dropbox link, we only need to update `prebuilt_index.py`

Comment (lintool) [reply docs/experiments-ltr-msmarco-passage.md:31]: > next time we change the dropbox link, we only need to update `prebuilt_index.py`

Correct!

**Comment by lintool** [comment]
Hey @stephaniewhoo at your branch, I ran:

```
$ python -m unittest integrations/sparse/test_ltr_msmarco_passage.py 
```

Got errors:

```
load queries
add feature
analyzed contents
text_unlemm text_unlemm
text_bert_tok text_bert_tok
IBM model Load takes 15.11 seconds
IBM model Load takes 31.98 seconds
IBM model Load takes 318.49 seconds
IBM model Load takes 57.30 seconds
Traceback (most recent call last):
  File "scripts/ltr_msmarco-passage/rerank_with_ltr_model.py", line 428, in <module>
    models = pickle.load(open(args.model+'/model.pkl', 'rb'))
  File "/anaconda3/envs/python36/lib/python3.6/site-packages/lightgbm/__init__.py", line 8, in <module>
    from .basic import Booster, Dataset, register_logger
  File "/anaconda3/envs/python36/lib/python3.6/site-packages/lightgbm/basic.py", line 95, in <module>
    _LIB = _load_lib()
  File "/anaconda3/envs/python36/lib/python3.6/site-packages/lightgbm/basic.py", line 86, in _load_lib
    lib = ctypes.cdll.LoadLibrary(lib_path[0])
  File "/anaconda3/envs/python36/lib/python3.6/ctypes/__init__.py", line 426, in LoadLibrary
    return self._dlltype(name)
  File "/anaconda3/envs/python36/lib/python3.6/ctypes/__init__.py", line 348, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(/anaconda3/envs/python36/lib/python3.6/site-packages/lightgbm/lib_lightgbm.so, 6): Library not loaded: /usr/local/opt/libomp/lib/libomp.dylib
  Referenced from: /anaconda3/envs/python36/lib/python3.6/site-packages/lightgbm/lib_lightgbm.so
  Reason: image not found
Traceback (most recent call last):
  File "tools/scripts/msmarco/msmarco_passage_eval.py", line 184, in <module>
    main()
  File "tools/scripts/msmarco/msmarco_passage_eval.py", line 173, in main
    metrics = compute_metrics_from_files(path_to_reference, path_to_candidate)
  File "tools/scripts/msmarco/msmarco_passage_eval.py", line 158, in compute_metrics_from_files
    qids_to_ranked_candidate_passages = load_candidate(path_to_candidate)
  File "tools/scripts/msmarco/msmarco_passage_eval.py", line 75, in load_candidate
    with open(path_to_candidate,'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'ltr_test/run.ltr.msmarco-passage.test.tsv'
E
```
Seems to be a lightgbm issue? For reference:

```
$ pip list | grep lightgbm
lightgbm                      3.2.1
```

**Comment by stephaniewhoo** [comment]
hrm, I have updated all my packages to [this](https://github.com/castorini/pyserini/issues/599#issuecomment-845886910), specifically:
```
$ pip list | egrep '(pyjnius|transformers|torch|sentencepiece|faiss-cpu|scikit-learn|lightgbm|spacy)\s'
faiss-cpu                     1.7.0
lightgbm                      3.2.1
pyjnius                       1.3.0
scikit-learn                  0.24.2
sentencepiece                 0.1.95
spacy                         3.0.6
torch                         1.8.1
transformers                  4.6.1
```

Running `python -m unittest integrations/sparse/test_ltr_msmarco_passage.py` on a brand newly cloned repo (no pre-download files effect the test), it still works ...
```
.
----------------------------------------------------------------------
Ran 1 test in 676.920s

OK
```

[this](https://github.com/microsoft/LightGBM/issues/1369) looks similar to your issue...

**Comment by lintool** [comment]
@stephaniewhoo ah... seems like a macOS issue. Will try to sort out...

**Comment by lintool** [comment]
Hi @stephaniewhoo I confirm that the issue was on my end - it was a macOS issue with lightgbm.

I confirm that the test case passes.

Let's go ahead and merge this in!

**Review by lintool — APPROVED** [review]

**Pull request merged**